src/libraries/System.Linq.Queryable/tests/TrimmingTests/System.Linq.Queryable.TrimmingTests.proj

```diff
@@ -2,7 +2,10 @@
   <Import Project="$([MSBuild]::GetPathOfFileAbove(Directory.Build.props))" />
 
   <ItemGroup>
-    <TestConsoleAppSourceFiles Include="QueryableTests.cs" />
+    <TestConsoleAppSourceFiles Include="QueryableTests.cs">
+      <!-- Queryable is not compatible with AOT -->
```

Comment:
This isn't good. The test is pretty simple, I'm surprised it doesn't work. Does it have any AOT warnings in it?

Isn't Queryable going to be needed with EF Core being on AOT? cc @roji @SamMonoRT

Reply:
The test has both AOT and trimming warnings on the `AsQueryable` call. We made some narrow happy path for trimming using DynamicDependency, but that doesn't work for AOT because the DynamicDependency refers to a generic method that we later need to MakeGenericMethod.

If queryable could be rewritten so that it does MakeGenericMethod when we still know the instantiation (I believe it matches the instantiation of the generic queryable method), it could take advantage of #99037 and be AOT safe for this narrow path as well.

Reply:
I don't have the full context here, but... If the issue is with AsQueryable() specifically, then EF should be fine (it shouldn't be required for EF queries). But if anything else related to Queryable/IQueryable is problematic, that may be something we need to look at...

Reply:
To provide more context, the exception at runtime with AOT is:

```
Unhandled exception. System.NotSupportedException: 'System.Linq.Enumerable.OrderByDescending[Program+Customer,System.Int32](System.Collections.Generic.IEnumerable`1[Program+Customer],System.Func`2[Program+Customer,System.Int32])' is missing native code. MethodInfo.MakeGenericMethod() is not compatible with AOT compilation. Inspect and fix AOT related warnings that were generated when the app was published. For more information see https://aka.ms/nativeaot-compatibility
   at System.Reflection.Runtime.MethodInfos.RuntimeNamedMethodInfo`1.GetUncachedMethodInvoker(RuntimeTypeInfo[], MemberInfo) + 0x5f
   at System.Reflection.Runtime.MethodInfos.RuntimeConstructedGenericMethodInfo.get_UncachedMethodInvoker() + 0x42
   at System.Reflection.Runtime.MethodInfos.RuntimeMethodInfo.get_MethodInvoker() + 0x54
   at System.Reflection.Runtime.MethodInfos.RuntimeNamedMethodInfo`1.MakeGenericMethod(Type[]) + 0x36a
   at System.Linq.EnumerableRewriter.<ArgsMatch>g__GetConstrutedGenericParameters|13_0(MethodInfo, Type[]) + 0x32
   at System.Linq.EnumerableRewriter.ArgsMatch(MethodInfo, ReadOnlyCollection`1, Type[]) + 0x1cc
   at System.Linq.EnumerableRewriter.<>c__DisplayClass11_0.<FindEnumerableMethodForQueryable>b__1(MethodInfo m) + 0x2c
   at System.Linq.Enumerable.IEnumerableWhereSelectIterator`2.ToArray() + 0x18e
   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1) + 0x5e
   at System.Linq.EnumerableRewriter.FindEnumerableMethodForQueryable(String, ReadOnlyCollection`1, Type[]) + 0x250
   at System.Linq.EnumerableRewriter.VisitMethodCall(MethodCallExpression) + 0x29e
   at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor) + 0x2e
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression) + 0x3b
   at System.Linq.Expressions.ExpressionVisitor.Visit(ReadOnlyCollection`1) + 0xa5
   at System.Linq.EnumerableRewriter.VisitMethodCall(MethodCallExpression) + 0x97
   at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor) + 0x2e
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression) + 0x3b
   at System.Linq.Expressions.ExpressionVisitor.Visit(ReadOnlyCollection`1) + 0xa5
   at System.Linq.EnumerableRewriter.VisitMethodCall(MethodCallExpression) + 0x97
   at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor) + 0x2e
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression) + 0x3b
   at System.Linq.Expressions.ExpressionVisitor.Visit(ReadOnlyCollection`1) + 0xa5
   at System.Linq.EnumerableRewriter.VisitMethodCall(MethodCallExpression) + 0x97
   at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor) + 0x2e
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression) + 0x3b
   at System.Linq.EnumerableExecutor`1.Execute() + 0x66
   at System.Linq.EnumerableQuery`1.System.Linq.IQueryProvider.Execute[TElement](Expression) + 0xd7
   at System.Linq.Queryable.Single[TSource](IQueryable`1 source) + 0x138
   at Program.Main(String[] args) + 0x2cc
```

We're hitting it on:

https://github.com/dotnet/runtime/blob/a45853c4751b5f532cdb38e3db5d4324b5ca878a/src/libraries/System.Linq.Queryable/tests/TrimmingTests/QueryableTests.cs#L18-L22

What I believe the test is testing is the `DynamicDependency` that was placed here:

https://github.com/dotnet/runtime/blob/a45853c4751b5f532cdb38e3db5d4324b5ca878a/src/libraries/System.Linq.Queryable/src/System/Linq/Queryable.cs#L417-L428

This is enough to keep the method, but we don't have a callable method body for it (we'd generate it over reference types only, but this needs Int32).

This does generate a trimming and AOT warnining, but the warning is on AsQueryable. If it's possible to enter this without IsQueryable, we probably have an illegal suppression somewhere or an analysis bug because there should be another warning.

Reply:
> This does generate a trimming and AOT warnining, but the warning is on AsQueryable. If it's possible to enter this without IsQueryable, we probably have an illegal suppression somewhere or an analysis bug because there should be another warning.

No, this looks okay. `EnumerableRewriter` (that is doing `MakeGeneric`) is marked RDC. This is called from `EnumerableQuery1` that is also marked RDC. And we instantiate `EnumerableQuery` in `AsQueryable` (the place that warns).

So the DynamicDependency is not actually for the sake of OrderByDescending, but for "OrderByDescending used together with AsQueryable". If it's not used together, the DynamicDependency probably just keeps some garbage we wouldn't reference otherwise.

The question is only whether we want this narrow happy path to work with AOT (despite hitting both AOT and trimming warnings).

Reply:
Looks like the narrow happy path was introduced in https://github.com/dotnet/runtime/pull/42603 and there's a bunch of discussion around it. Seems like I was already skeptical about it, but I don't remember any of that.

If the call was on me, I'd lean towards just deleting the DynamicDependency and these tests.

Reply:
Oh, I see, this was added in response to a bug report from appmodels that just disable trimming warnings and hope for the best. It probably works often enough that people are mostly satisfied (to the extent that they can be satisfied with unreliable trimming).

We could probably do something to make this work for AOT as well. It will not be pretty. The compiler needs to see a call to MakeGeneric with the Queryable method body where we still know all the T's and U's. Opinions?

Reply:
Can we at least log an issue for this, and stick the issue number here? Once we do that, I'm good with merging this.

Reply:
```suggestion
      <!-- https://github.com/dotnet/runtime/issues/101903 -->
```